S3/FileLists.py

```diff
@@ -60,34 +60,37 @@ def handle_exclude_include_walk_dir(root, dirname):
     Should this root/dirname directory be excluded? (otherwise included by default)
     Exclude dir matches in the current directory
     This prevents us from recursing down trees we know we want to ignore
-    return True for including, and False for excluding
+    return True for excluding, and False for including
     '''
     cfg = Config()
+    # python versions end their patterns (from globs) differently, test for different styles
+    directory_patterns = [u'\\/$', u'\\/\\Z(?ms)', u'/)$', u'/)\\Z']
+
     d = os.path.join(root, dirname, '')
     debug(u"CHECK: %r" % d)
     excluded = False
     for r in cfg.exclude:
-        # python versions end their patterns (from globs) differently, test for both styles.
-        if not (r.pattern.endswith(u'\\/$') or r.pattern.endswith(u'\\/\\Z(?ms)')): continue # we only check for directory patterns here
+        if not any(r.pattern.endswith(dp) for dp in directory_patterns): continue # we only check for directory patterns here
+        debug(u"EXCL-TEST: %s ~ %s" % (d, r.pattern))
         if r.search(d):
             excluded = True
             debug(u"EXCL-MATCH: '%s'" % (cfg.debug_exclude[r]))
             break
     if excluded:
         ## No need to check for --include if not excluded
         for r in cfg.include:
-            # python versions end their patterns (from globs) differently, test for both styles.
-            if not (r.pattern.endswith(u'\\/$') or r.pattern.endswith(u'\\/\\Z(?ms)')): continue # we only check for directory patterns here
+            if not any(r.pattern.endswith(dp) for dp in directory_patterns): continue # we only check for directory patterns here
             debug(u"INCL-TEST: %s ~ %s" % (d, r.pattern))
             if r.search(d):
                 excluded = False
                 debug(u"INCL-MATCH: '%s'" % (cfg.debug_include[r]))
                 break
     if excluded:
         ## Still excluded - ok, action it
-        debug(u"EXCLUDE: %r" % d)
+        ## Use relative name to easily check this in tests
+        debug(u"EXCLUDE: %r" % dirname)
```

Comment:
Interesting idea, but still I think that it is better to have the full name.
Because if debugging the behavior of s3cmd, it might be easier to understand an issue if we know where is the folder that is excluded.

Reply:
> it is better to have the full name

Yes, it's reasonable. 

But then I'll need to fix my "Exclude directory" test, because EXCLUDE debug messages will contain absolute paths that depend on local environment. I can use regexps (must_find_re/must_not_find_re) with broad patterns (`.*`) instead of simple string comparison (must_find/must_not_find), this way doesn't look pretty for me, but if there's no other way, then I'll do it. What do you think?

Reply:
It is better to have the code more helpful and the test less good than the opposite.

For the test, the following regexp should be enough, no?
'testsuite/demo/dir1$'
otherwise .* is ok without issue.

Reply:
ok, got it
also, 1) I was wrong about full path in debug messages, actually I don't need any regexp in my test, 2) found and fixed a bug in my test (wrong file names in must_not_find list)

Now everything looks fine